### PR TITLE
planner/cascades: support parameterized transformation/implementation rules

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 

Before this PR, cascades optimizer uses all of the transformation/implementation rules. This is not friendly to unit tests. Since we usually only want to check the specific transformation rule, and we don't want the unit tests affected by the new added rules, we must be able to choose the transformation rule map for a specific unit test.

Furthermore, if we want to support multi-level optimization, we also have to choose different transformation rule map for different optmization level.

### Check List 

Tests 

No tests. But will change the unit tests after other PRs of cascades merged.

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -40,13 +40,13 @@ type Transformation interface {
 	OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error)
 }
 
-// GetTransformationRules gets the all the candidate transformation rules based
-// on the logical plan node.
-func GetTransformationRules(node plannercore.LogicalPlan) []Transformation {
-	return transformationMap[memo.GetOperand(node)]
+// GetTransformationRules gets the all the candidate transformation rules of the optimizer
+// based on the logical plan node.
+func (opt *Optimizer) GetTransformationRules(node plannercore.LogicalPlan) []Transformation {
```
Comment (zz-jason) [inline planner/cascades/transformation_rules.go]: How about putting all the functions bond to the `Optimizer` struct together into the `optimize.go` source file?

File: planner/cascades/optimize.go [hunk 2]
```diff
@@ -23,6 +23,30 @@ import (
...
+func NewDefaultOptimizer() *Optimizer {
+	return &Optimizer{
+		transformationRuleMap: defaultTransformationMap,
+		implementationRuleMap: defaultImplementationMap,
+	}
+}
+
+// NewOptimizerWithTrans returns a cascades optimizer which uses user defined
+// transformation rule map. This will be frequently used in unit tests.
+func NewOptimizerWithTrans(trans map[memo.Operand][]Transformation) *Optimizer {
```
Comment (zz-jason) [inline planner/cascades/optimize.go]: How about adding the following two functions for the `Optimizer` struct:
```go
func (opt *Optimizer) ResetTransformationRules(map[memo.Operand][]Transformation) { ... }
func (opt *Optimizer) ResetImplementationRules(map[memo.Operand][]ImplementationRule) { ... }
```

Comment (francis0407) [reply planner/cascades/optimize.go]: Looks good.
It's convenient for unit tests, then we only need one `Optimizer` in a test suite, and we can reset the rules in each unit tests.

File: planner/cascades/transformation_rules.go (hunk 1 above)
Comment (francis0407) [reply planner/cascades/transformation_rules.go]: addressed

File: planner/optimize.go [hunk 3]
```diff
@@ -105,7 +105,7 @@ func optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 
 	// Handle the logical plan statement, use cascades planner if enabled.
 	if sctx.GetSessionVars().EnableCascadesPlanner {
-		return cascades.FindBestPlan(sctx, logic)
+		return cascades.NewOptimizer().FindBestPlan(sctx, logic)
```
Comment (zz-jason) [inline planner/optimize.go]: Can we create a singleton for it, for example `defaultOptimizer`? Then we can use it globally, don't need to create a new `Optimizer` object for each query.

Comment (francis0407) [reply planner/optimize.go]: Sure, I created it in `planner/cascades/optimizer`.

**Pull request merged**